akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -401,6 +480,15 @@ private[remote] final class EnvelopeBuffer(val byteBuffer: ByteBuffer) {
     }
   }
 
+  @tailrec private def writeByteString(bs: ByteString): Unit = {
+    bs match {
+      case cs: ByteString1C ⇒
+        byteBuffer.putInt(bs.length)
+        cs.copyToBuffer(byteBuffer)
+      case _ ⇒ writeByteString(bs.compact)
```

Comment:
Is it really faster to compact the `ByteString` instead of just doing `bs.length` and `bs.copyToBuffer`? Isn't `ByteString` optimized to avoid unnecessary copies?